docs/modules.md

```diff
@@ -329,6 +329,96 @@ For example, if the user `@alice:example.org` is passed to this method, and the
+their Synapse installation with an external authentication system. The callbacks can be
+registered by using the Module API's `register_password_auth_provider_callbacks` method.
+
+To register authentication checkers, the module should register:
+
+```
+ auth_checkers: Dict[Tuple[str,Tuple], CHECK_AUTH_CALLBACK]
+```
+
+A dict mapping from tuples of a login type identifier (such as `m.login.password`) and a
+tuple of field names (such as `("password", "secret_thing")`) to authentication checking
+methods.
+
+The login type and field names are the things that are provided by the user in their submission
+to the `/login` API. 
+
+For example, if a module implements authentication checkers for two different login types: 
+-  `com.example.custom_login` 
+    - Expects `secret1` and `secret2` fields to be sent to `/login`
+    - Is checked by the method: `self.custom_login_check`
+- `m.login.password`
+    - Expects a `password` field to be sent to `/login`
+    - Is checked by the method: `self.password_checker` 
+    
+Then it should register the following dict:
+
+```python
+{ 
+    ("com.example.custom_login", ("secret1", "secret2")): self.check_custom_login,
+    ("m.login.password", ("password",)): self.password_checker,
+}
+```
+
+An authentication checking method should be of the following form:
+
+```python
+async def check_auth(
+        username: str,
+        login_type: str,
+        login_dict: JsonDict
+) -> Optional[Tuple[str, Optional[Callable]]]
+```
+
+It is passed the user field provided by the client (which might not be in `@username:server` form), 
+the login type, and a dictionary of login secrets passed by the client.
+
+On a failure it should return None. On a success it should return a (user_id, callback) tuple
+where user_id is a str containing the user's matrix id (i.e. in `@username:server` form) and the 
+callback is a method to be called with the result from the `/login` call (see the 
+<a href="https://spec.matrix.org/unstable/client-server-api/#login">spec</a> 
+for details). This callback can be None.
+
+Aditionally the module can register:
+
+```python
+async def check_3pid_auth(
+    medium: str, 
+    address: str,
+    password: str,
+) -> Optional[Tuple[str, Optional[Callable]]]
+```
+
+This  is called when a user attempts to register or log in with a third party identifier,
+such as email. It is passed the medium (eg. "email"), an address (eg. "jdoe@example.com")
+and the user's password.
+
+On a failure it should return None. On a success it should return a (user_id, callback) tuple
+where user_id is a str containing the user's matrix id (i.e. in `@username:server` form) and the 
+callback is a method to be called with the result from the `/login` call (see the 
+<a href="https://spec.matrix.org/unstable/client-server-api/#login">spec</a> 
+for details). This callback can be None.
+
+```python
+async def on_logged_out(
+        user_id: str,
+        device_id: str,
+        access_token: str,
+) -> None
```

Comment:
The indentation looks off. Also it looks like this should be able to fit on a single line.